### ISSUE TITLE
Rejected privacy consent 5 times only

### ISSUE THREAD

**Issue by suchyjan** [Songkeys/Translateer#15]
Hello Song,

I have encountered an interesting issue that I would like to discuss with you. When I set `PAGE_COUNT` to 15 and start the Translateer, the initialization process appears as follows:

export PAGE_COUNT=15
..
Initializing pages...
No privacy consent
No privacy consent
No privacy consent
No privacy consent
No privacy consent
No privacy consent
No privacy consent
No privacy consent
No privacy consent
No privacy consent
Rejected privacy consent
Rejected privacy consent
Rejected privacy consent
Rejected privacy consent
Rejected privacy consent
Ready
..

However, when I set `PAGE_COUNT` to 5, the initialization is as follows:

export PAGE_COUNT=5
..
Initializing pages...
Rejected privacy consent
Rejected privacy consent
Rejected privacy consent
Rejected privacy consent
Rejected privacy consent
Ready

It seems that in the first example, the privacy consent is not rejected entirely (only five rejections are fixed, but not the others). After some time, possibly around one hour (when pages are reloaded), the following error occasionally occurs when calling translation:

res":{"statusCode":500},"err":{"type":"Error","message":"Error: failed to find an element matching selector \"button[aria-label='Clear source text']\"","stack":"Error: Error: failed to find an element matching selector \"button[aria-label='Clear source text']\"\n    at CDPElementHandle.$eval (/home/owle/app/translateer/node_modules/puppeteer-core/lib/cjs/puppeteer/common/ElementHandle.js:140:19)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"},"msg":"Error: failed to find an element matching selector \"button[aria-label='Clear source text']\""}

Upon restarting the Translateer, the error sometimes disappears, but at other times, it persists.

I am curious whether the rejection of consent is flexible based on the `PAGE_COUNT` variable or if it is fixed. This variation could potentially be the source of the issue.

Thank you for your review and assistance in resolving this matter.

Best regards,

Jan

**Comment by songkeys** [comment]
Hi, thank you for this report.

I'm currently not in the EU, so I'm unable to reproduce this.

Does it always reject 5 times only? The code currently waits for 1000ms for the privacy consent to appear and then rejects it. You can find the code here: [https://github.com/Songkeys/Translateer/blob/2d24c98731313bcee0a28f68748c9f19b2d3dda9/src/browser/pagepool.ts#L75](https://github.com/Songkeys/Translateer/blob/2d24c98731313bcee0a28f68748c9f19b2d3dda9/src/browser/pagepool.ts#L75)

I think this behavior looks like a network issue because when too many pages are opened at the same time, some might get stuck, and the privacy consent might appear after more than 1000ms. But in this case, the results should vary instead of exactly 5 times working versus 10 times not working every time if `PAGE_COUNT=15`.

To help debug this issue, could you please change the line of code above to a longer time value like 5000ms or something else? Also, changing `PAGE_COUNT` to other values and observing the behavior might help. To understand exactly what's going on, you could run `npm run debug` on your local machine to have the browser open and visually see what's happening.

If this is difficult for you, don't worry - I'll check back once I have an EU VPN during the following weeks.

**Comment by suchyjan** [comment]
Hi, Song,

It's interesting, but I upgraded Chromium to a newer version on the Ubuntu server where this was happening and now it won't acknowledge Consent even once and yet everything works. I did a us run debug with the X server configured to see the window on my macOS and on Ubuntu it shows the google service now without the consent dialog - I tried whaismyip.com from the browser and the IP was from Helsinki, Finland (i.e. EU).
It seems as if the installation of Chromium has had an effect.. 

Ubuntu:
Version 117.0.5938.92 (Official Build) snap (64-bit)
- Does not show Consent from Finland

-

On macOS, I installed the latest version of Chromium from homebrew and got the developer version. However, running there detects the IP address of the Czech Republic, which is ok and it is also in the EU and consent is displayed correctly as it should.

MacOS:
Version 119.0.6032.0 (Developer build) (arm64)
- Displays Consent from EU always

I was originally getting the error on Ubuntu, but after installing Chrome it no longer shows up due to Consent not even displaying to the user..

On macOS I tried to simulate multiple windows PAGE_COUNT=55 and indeed for some windows consent failed to close, even though it was displayed according to debug in all.. This is how it behaved originally on Ubuntu as well. However, I have tested the timeout you mentioned in such a case and it is indeed because consent fails to close in the given timeout with multiple windows open. So I increased it to 2000 and already all consents are confirmed OK.



On slower machines this can be a problem, sometimes the total call reaches the 30sec limit and crashes on puppeteer error:

..app/translateer/node_modules/puppeteer-core/lib/cjs/puppeteer/common/LifecycleWatcher.js:174
    return new Errors_js_1.TimeoutError(errorMessage);
           ^
TimeoutError: Navigation timeout of 30000 ms exceeded
    at LifecycleWatcher._LifecycleWatcher_createTimeoutPromise (..app/translateer/node_modules/puppeteer-core/lib/cjs/puppeteer/common/LifecycleWatcher.js:174:12)

This is of course for another / setting timeouts for Puppeteer.. and to be expected..

I think this solves the problem and can be closed if you agree.

I was a bit surprised that consent stopped being needed for the EU location (Finland) after installing the new Chromium, but no more to investigate for now..

Thanks,
Jan
